File: pybind_adv/src/pybind/cpp_thread_callback/tick_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# pylint: disable=missing-docstring, import-error
import functools
import time

import tick_pb as tick


def _main():
  t = tick.TaskRunner()

  time.sleep(1)
  def tick_event1():
      print("one")
  def tick_event2():
      print("two")

  print("pushing one")
  t.push_task(tick.Task(lambda:print("one")))
  print("pushing two")
  t.push_task(tick.Task(lambda:print("two")))

  print("sleeping")
  time.sleep(4)


if __name__ == "__main__":
  _main()

File: pybind_adv/src/cpp/cpp_thread_callback/tick_test.cc
#include <iostream>
#include <chrono>
#include <thread>

#include "tick.h"

int main(int argc, char const *argv[]) {
  (void)argc;
  (void)argv;

  {
     TaskRunner tr;

     Task t1([]() {
           std::cout << "one" << std::endl;
           });
     Task t2([]() {
           std::cout << "two" << std::endl;
           });

     tr.PushTask(t1);
     tr.PushTask(t2);


     using namespace std::chrono_literals;
     std::this_thread::sleep_for(4s);
     std::cout << "done\n";
  }

  return 0;
}
